File: Medium/2026-08-19-1386-Cinema-Seat-Allocation/solution.java
import java.util.HashSet;
import java.util.Set;

public class Solution {
    public int maxNumberOfFamilies(int n, int[][] reservedSeats) {
        Set<Integer> reserved = new HashSet<>();
        for (int[] seat : reservedSeats) {
            reserved.add(seat[0] * 10 + seat[1]);
        }

        int maxGroups = 0;

        for (int row = 1; row <= n; row++) {
            int available = 0;
            if (!reserved.contains(row * 10 + 2) && !reserved.contains(row * 10 + 3) && 
                !reserved.contains(row * 10 + 4) && !reserved.contains(row * 10 + 5)) {
                available++;
            }
            if (!reserved.contains(row * 10 + 4) && !reserved.contains(row * 10 + 5) && 
                !reserved.contains(row * 10 + 6) && !reserved.contains(row * 10 + 7)) {
                available++;
            }
            if (!reserved.contains(row * 10 + 6) && !reserved.contains(row * 10 + 7) && 
                !reserved.contains(row * 10 + 8) && !reserved.contains(row * 10 + 9)) {
                available++;
            }
            maxGroups += available;
        }

        return maxGroups;
    }
}

File: Medium/2026-08-19-1386-Cinema-Seat-Allocation/solution.js
var maxNumberOfFamilies = function(n, reservedSeats) {
    const reserved = new Map();
    
    for (const [row, seat] of reservedSeats) {
        if (!reserved.has(row)) {
            reserved.set(row, new Set());
        }
        reserved.get(row).add(seat);
    }
    
    let maxGroups = 0;

    for (let i = 1; i <= n; i++) {
        const reservedSeatsInRow = reserved.get(i) || new Set();
        let count = 0;

        if (!reservedSeatsInRow.has(2) && !reservedSeatsInRow.has(3) && !reservedSeatsInRow.has(4) && !reservedSeatsInRow.has(5)) {
            count++;
        }
        if (!reservedSeatsInRow.has(4) && !reservedSeatsInRow.has(5) && !reservedSeatsInRow.has(6) && !reservedSeatsInRow.has(7)) {
            count++;
        }
        if (!reservedSeatsInRow.has(6) && !reservedSeatsInRow.has(7) && !reservedSeatsInRow.has(8) && !reservedSeatsInRow.has(9)) {
            count++;
        }

        maxGroups += count === 0 ? 0 : count === 2 ? 2 : 1;
    }

    return maxGroups;
};